File: client/mobile_flutter/lib/features/chat/domain/entities/conversation.dart
import 'package:equatable/equatable.dart';

import '../../../auth/domain/entities/user.dart';
import 'message.dart';

/// copyWith 中用于区分 null 和未提供的哨兵值
const _sentinel = Object();

/// 会话类型枚举
enum ConversationType { private, group, channel }

/// 会话实体
class Conversation extends Equatable {
  const Conversation({
    required this.id,
    required this.type,
    required this.members,
    required this.createdAt,
    this.name,
    this.creatorId,
    this.lastMessage,
    this.unreadCount = 0,
  });

  final String id;
  final ConversationType type;
  final List<User> members;
  final DateTime createdAt;
  final String? name;
  final String? creatorId;
  final Message? lastMessage;
  final int unreadCount;

  Conversation copyWith({
    String? id,
    ConversationType? type,
    List<User>? members,
    DateTime? createdAt,
    Object? name = _sentinel,
    Object? creatorId = _sentinel,
    Object? lastMessage = _sentinel,
    int? unreadCount,
  }) {
    return Conversation(
      id: id ?? this.id,
      type: type ?? this.type,
      members: members ?? this.members,
      createdAt: createdAt ?? this.createdAt,
      name: name == _sentinel ? this.name : name as String?,
      creatorId: creatorId == _sentinel ? this.creatorId : creatorId as String?,
      lastMessage: lastMessage == _sentinel ? this.lastMessage : lastMessage as Message?,
      unreadCount: unreadCount ?? this.unreadCount,
    );
  }

  @override
  List<Object?> get props => [
        id,
        type,
        members,
        createdAt,
        name,
        creatorId,
        lastMessage,
        unreadCount,
      ];
}

File: service/chat_gin/internal/model/conversation.go
package model

import (
	"time"

	"github.com/google/uuid"
)

// ConversationType 会话类型枚举
type ConversationType string

const (
	ConversationTypePrivate ConversationType = "private" // 私聊（一对一）
	ConversationTypeGroup   ConversationType = "group"   // 群聊（多人）
	ConversationTypeChannel ConversationType = "channel" // 频道（广播模式）
)

// Conversation 会话实体模型
type Conversation struct {
	ID        uuid.UUID        `json:"id" gorm:"type:uuid;primary_key;default:gen_random_uuid()"`
	Type      ConversationType `json:"type" gorm:"type:varchar(20);not null"`
	Name      string           `json:"name" gorm:"type:varchar(100)"`
	CreatorID uuid.UUID        `json:"creator_id" gorm:"type:uuid;not null"`
	CreatedAt time.Time        `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time        `json:"updated_at" gorm:"autoUpdateTime"`

	// 关联关系
	Members     []ConversationMember `json:"members,omitempty" gorm:"foreignKey:ConversationID"`
	LastMessage *Message             `json:"last_message,omitempty" gorm:"-"` // 最后一条消息（不存储在数据库）
	UnreadCount int                  `json:"unread_count" gorm:"-"`           // 未读消息数（不存储在数据库，动态计算）
}

// TableName 返回会话表名
func (Conversation) TableName() string {
	return "chat_conversations"
}

// ConversationMember 会话成员实体模型
type ConversationMember struct {
	ConversationID uuid.UUID `json:"-" gorm:"type:uuid;primaryKey"`
	UserID         uuid.UUID `json:"id" gorm:"type:uuid;primaryKey"`
	Role           string    `json:"role" gorm:"type:varchar(20);default:'member'"`
	JoinedAt       time.Time `json:"-" gorm:"autoCreateTime"`
	LastReadAt     time.Time `json:"last_read_at" gorm:"default:null"`

	// 用户信息（从认证服务获取，不存储在数据库）
	Username    string  `json:"username,omitempty" gorm:"-"`
	Email       string  `json:"email,omitempty" gorm:"-"`
	DisplayName *string `json:"display_name,omitempty" gorm:"-"`
	AvatarURL   *string `json:"avatar_url,omitempty" gorm:"-"`
}

// TableName 返回会话成员表名
func (ConversationMember) TableName() string {
	return "chat_conversation_members"
}

// 成员角色常量
const (
	MemberRoleOwner  = "owner"  // 群主
	MemberRoleAdmin  = "admin"  // 管理员
	MemberRoleMember = "member" // 普通成员
)

// GetMemberIDs 获取所有成员的用户ID列表
func (c *Conversation) GetMemberIDs() []uuid.UUID {
	ids := make([]uuid.UUID, len(c.Members))
	for i, m := range c.Members {
		ids[i] = m.UserID
	}
	return ids
}

// HasMember 检查用户是否是会话成员
func (c *Conversation) HasMember(userID uuid.UUID) bool {
	for _, m := range c.Members {
		if m.UserID == userID {
			return true
		}
	}
	return false
}
